### PR TITLE
[AIRFLOW-5072] gcs_hook's download() method should download only once

### PR DESCRIPTION
**Pull request by tkaymak**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://jira.apache.org/jira/browse/AIRFLOW-5072) issues and references them in the PR title.
  - https://jira.apache.org/jira/browse/AIRFLOW-5072
### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: missing else inserted

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

**Comment by OmerJog** [comment]
You have pylint error to fix:
`airflow/contrib/hooks/gcs_hook.py:157:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)`

**Comment by tkaymak** [comment]
Closing and re-opening PR as Travis seems to have issues which are not related to the code

**Comment by tkaymak** [comment]
what should I do? the build is failing for no obvious reason (timeout)

**Comment by potiuk** [comment]
You can also ask on Slack to restart it. Hopefully when we revamp kubernetes tests (in progress) and switch to GitLab CI (as well) it will not be happening that often.

File: airflow/contrib/hooks/gcs_hook.py
Comment (mik-laj) [inline airflow/contrib/hooks/gcs_hook.py:183]: I think that the value of this result should be consistent. In both cases it should be the content/stream of the file.

Comment (tkaymak) [reply airflow/contrib/hooks/gcs_hook.py:183]: @mik-laj that is breaking the whole thing as when the blob is huge it eats up all the memory and crashes.
As a user when I supply a filename to download to a file I don't want the content of that file being passed back to me. My expectation is that it's written to that filename and done with it.

Comment (mik-laj) [reply airflow/contrib/hooks/gcs_hook.py:183]: Can you present your assumptions in the docstring?  I think it's worth describing this behavior and the reason for this behavior, not another behavior.

Comment (tkaymak) [reply airflow/contrib/hooks/gcs_hook.py:183]: @mik-laj I've updated the docstring

Comment (tkaymak) [reply airflow/contrib/hooks/gcs_hook.py:183]: @mik-laj so is this a yes? :) or should we wait for another review?

Comment (mik-laj) [reply airflow/contrib/hooks/gcs_hook.py:183]: I'm on vacation untill Friday. 🇫🇷 Please wait. I will return to this when I return.

Comment (tkaymak) [reply airflow/contrib/hooks/gcs_hook.py:183]: Enjoy! 🥖

Comment (turbaszek) [reply airflow/contrib/hooks/gcs_hook.py:183]: Information about this change should be propagated to GCS download operator:
https://github.com/apache/airflow/blob/ec7c67ff95e012d5609a56d358fc6c29d01d9ff6/airflow/contrib/operators/gcs_download_operator.py#L28-L41

Comment (tkaymak) [reply airflow/contrib/hooks/gcs_hook.py:183]: @mik-laj / @nuclearpinguin done

File: airflow/contrib/operators/gcs_download_operator.py
Comment (ashb) [inline airflow/contrib/operators/gcs_download_operator.py]: This isn't quite true for an operator as operators don't have a return value, so it's the store_to_xcom_key that controls the behaviour.

As a result of your change in the hook `store_to_xcom_key` and `filename` are now mutually exclusive, which makes this a breaking change to the operator behaviour. 

So either you need to add a note to UPDATING.md about this change (and ensure that either store_to_xcom_key is True and filename is empty) or make the operator load the contents form disk when store_to_xcom_key is True and filename is not empty.

Comment (tkaymak) [reply airflow/contrib/operators/gcs_download_operator.py]: @ashb Thank you for your comment, I've changed the behaviour, updated the docstring and included the info into the UPDATING.md

File: UPDATING.md
Comment (ashb) [inline UPDATING.md]:
```suggestion
The `GoogleCloudStorageDownloadOperator` can either write to a supplied `filename` or return the content of a file via xcom through `store_to_xcom_key` - both options are mutually exclusive.
```

File: tests/contrib/hooks/test_gcs_hook.py
Comment (ashb) [inline tests/contrib/hooks/test_gcs_hook.py:561]: Does this test now also leave `test_file` file hanging around? (I think it does.) We should tidy that up with a `try`/`finally`.

File: airflow/contrib/operators/gcs_download_operator.py
Comment (ashb) [inline airflow/contrib/operators/gcs_download_operator.py:37]:
```suggestion

    If a filename is supplied, it writes the file to the specified location, alternatively one can
```

Need a blank line between the "summary" and body of the doc.

File: airflow/contrib/operators/gcs_download_operator.py
Comment (ashb) [inline airflow/contrib/operators/gcs_download_operator.py]:
```suggestion
    set the ``store_to_xcom_key`` parameter to True push the file content into xcom. When the file size
```

File: airflow/contrib/hooks/gcs_hook.py
Comment (ashb) [inline airflow/contrib/hooks/gcs_hook.py:162]:
```suggestion
        Downloads a file from Google Cloud Storage.
        
```

**Review by ashb — CHANGES_REQUESTED** [review]

**Review by ashb — APPROVED** [review]

**Comment by potiuk** [comment]
That was the intermittent kubernetes test that failed. I merge it now without waiting as it blocks some of our refactorings to GCP.

**Pull request merged**
